### PR TITLE
fix(ontology): disable node select when searching so users don't select hidden terms

### PR DESCRIPTION
**Pull request by mswertz**
in part https://github.com/molgenis/GCC/issues/819
(other half in #4539)

to test:
- when using ontology select in forms and filters it will remove the 'node' selector when searching
- see issue that was mentioned

Spec based on discussion
- [x] if search matches only a leaf and not its parent nodes, we show only the leaf plus its path. Node can not be selected
- [x] if search match only a node and not its child nodes, we show the node and then all its children without search filter

todo:
- [ ] check edge case when the node is found but none of its children, how should it then behave, is select all then ok?
- [ ] test if this leads to incomplete data entry
- [ ] check if there are other applications affected, e.g. directory and if that is desired

N.B. a next step would be to also enable a special ontology type (categorical) that would have node selection always disabled. See #ontology

### CODE REVIEW

File: apps/molgenis-components/src/components/forms/InputOntology.vue [hunk 1]
```diff
@@ -87,6 +87,7 @@
           :key="key"
           :terms="rootTerms"
           :isMultiSelect="isMultiSelect"
+          :enableSelectNodes="!search"
```
Comment (chinook25) [inline apps/molgenis-components/src/components/forms/InputOntology.vue:90]: ===

**Review by chinook25 — CHANGES_REQUESTED** [review]
I would like to see an example added to the component docs.

**Review by hslh — CHANGES_REQUESTED** [review]
It currently only disables the top-level node, is this the intended behaviour? I'd expect all non-matching non-leaf nodes to be disabled. 
Screenshot 2024-12-04 at 14 32 13
The unintentional selection of hidden children not matching the search criteria is still possible:
Screenshot 2024-12-04 at 14 32 25
Screenshot 2024-12-04 at 14 32 36

The behaviour for filters is different, there it doesn't select hidden children:
Screenshot 2024-12-04 at 14 31 18
Screenshot 2024-12-04 at 14 31 48
Screenshot 2024-12-04 at 14 31 37

File: apps/molgenis-components/src/components/forms/InputOntology.vue (hunk 1 above)
Comment (mswertz) [reply apps/molgenis-components/src/components/forms/InputOntology.vue:90]: huh? that doesn't make sense here this is a component assignment.

**Comment by mswertz** [comment]
good point @hslh. Added a change that should also apply this for the subtree.

note that this fix only applies to the data entry forms, not the catalogue. It was unclear to me if that was also desired. That should then need seperate fix.

**Review by hslh — COMMENTED** [review]
The input form looks better now, but I'm not sure about non-leaf nodes which do match the search criteria, e.g. you can't select 'I Certain infectious and parasitic diseases', even though it fits the search of 'parasitic'. 

Screenshot 2024-12-10 at 11 14 46

**Comment by hslh** [comment]
For the filters in the catalogue I'm not sure whether it's an improvement either. The same matter as for the input forms applies: non-leaf nodes matching the search criteria aren't available for selection, and I think they should be. 

The other problem with the catalogue filters is that they don't unfold when searching: 
Screenshot 2024-12-10 at 11 11 20

If they could unfold to show all the matching nodes, that'd make it better, like so:
Screenshot 2024-12-10 at 11 11 29

File: apps/molgenis-components/src/components/forms/InputOntology.vue (hunk 1 above)
Comment (chinook25) [reply apps/molgenis-components/src/components/forms/InputOntology.vue:90]: Wait, did the code change? This wasn't where I thought I put this comment 😅

File: apps/nuxt3-ssr/components/filter/Ontology.vue [hunk 2]
```diff
@@ -7,6 +7,7 @@ const props = withDefaults(
   defineProps<{
     tableId: string;
     filter: Record<String, Filter>;
+    enableSelectNodes: boolean;
```
Comment (chinook25) [inline apps/nuxt3-ssr/components/filter/Ontology.vue:10]: need to be optional, otherwise the default doesn't make sense here

File: apps/tailwind-components/components/input/Tree.vue [hunk 3]
```diff
@@ -9,12 +9,14 @@ const props = withDefaults(
     modelValue: string[];
     isMultiSelect?: boolean;
     expandSelected?: boolean;
+    enableSelectNodes: boolean;
```
Comment (chinook25) [inline apps/tailwind-components/components/input/Tree.vue:12]: optional

File: apps/tailwind-components/components/input/Tree.vue [hunk 4]
```diff
@@ -144,6 +146,7 @@ function handleChildSelect(selected: string[], parent: ITreeNode) {
           class="flex justify-center items-start hover:cursor-pointer"
         >
           <InputCheckboxIcon
+            v-if="enableSelectNodes || node.children?.length == 0"
```
Comment (chinook25) [inline apps/tailwind-components/components/input/Tree.vue:149]: ===

**Review by chinook25 — CHANGES_REQUESTED** [review]
It breaks if (on master it works): 

1. go to http://localhost:5173/#/component/InputOntology
2. in the first example select something, 
3. deselect it again
4. see console error
5. the input is broken now

**Comment by mswertz** [comment]
Outcome of discussion:
1. if search matches only a leaf and not its parent nodes, we show only the leaf plus its path. Node can not be selected
2. if search match only a node and not its child nodes, we show the node and then all its children without search filter
3. if search matches both a node and a child, we show the node with the leaves filtered. Node cannot be selected
  - bonus feature: for such a node we can click a 'show all' and then the behavior of '2' is used

**Comment by hslh** [comment]
Desired behaviour changed after testing, is now:
 
 (1) if search matches only a leaf and not its parent nodes, we show only the leaf plus its path. Node can not be selected.  (2) if search matches a node, we show the node and then all its children without search filter (even when search would narrow filter of the children).

**Review by hslh — DISMISSED** [review]
I think the current behaviour is good. It's less advanced than the three-way behaviour we decided on originally, but now I think it's easier to understand for the user what's going on. 

The behaviour of unfolding the tree to show all search results seems inconsistent, however. When I search for 'thoracic', everything stays folded, when I search for 'lingual', all matching results are expanded. Is this based on the number of results?
Screenshot 2024-12-16 at 10 23 29
Screenshot 2024-12-16 at 10 24 10

**Comment by mswertz** [comment]
> Is this based on the number of results?

correct. In the catalogue we don't do that. What do you prefer?

**Comment by hslh** [comment]
> > Is this based on the number of results?
> 
> correct. In the catalogue we don't do that. What do you prefer?

Keeping it like this is fine, you don't want the whole ICD-10 to expand after you type a single letter

**Review by chinook25 — CHANGES_REQUESTED** [review]
Tested on https://preview-emx2-pr-4538.dev.molgenis.org/apps/molgenis-components/#/component/InputOntology
Still found the following bug:  deselecting a node/leaf results in error: 't is null'

**Comment by mswertz** [comment]
> Tested on https://preview-emx2-pr-4538.dev.molgenis.org/apps/molgenis-components/#/component/InputOntology
> Still found the following bug: deselecting a node/leaf results in error: 't is null'

Interestingly when I run locally I also got this error on master

**Comment by mswertz** [comment]
I have moved most of these changes to https://github.com/molgenis/molgenis-emx2/pull/4539

**Review by chinook25 — APPROVED** [review]
